decisions/0015-separate-library-per-product.md

```diff
@@ -0,0 +1,72 @@
+---
+parent: Decisions
+nav_order: 15
+adr: "0015"
+status: In progress
+date: 2022-11-03
+tags: [clients]
+---
+
+# Separate library per product
+
+## Context and Problem Statement
+
+Our `clients` repository currently stores much common logic inside of the `libs/common` package,
+which is a library that is consumed by all of our clients. However not all parts of the package are
+actually used by all of our clients which makes our bundles unecessarily large. Another issue with
+this arrangement is the increased cognitive load for the different pods that have to filter out the
+parts of the package that don't concern their work.
+
+## Considered Options
+
+- **Continue as is** - Place all services, models, views, etc in `libs/common`.
+- **Put services that are only used in 1 client inside that client** - This has the benefit of not
+  bloating the common package but will start "polluting" our frontends with request/response models
+  that we're trying to hide behind services. Cognitive load is also not particularly improved.
+  Services that are used in more than 1 client would automatically end up in `common`.
+- **Create new libs for each product** - This involves splitting our common code into `common`,
+  `password-manager`, `admin-console` and `secrets-manager` library packages. It has all the benfits
+  of the `common` package such as being able to keep internal request/response models while avoiding
+  bloat and lowering cognitive load. It also continues to encourage separation of UI and domain
+  knowledge. The code inside the packages would be organized by feature.
+- **Create new libs for each feature** - Shares many similarities with the option above, but instead
+  splits the common code directly into features. Just like above, this has all the benfits of the
+  `common` package such as being able to keep internal request/response models while avoiding bloat
+  and lowering cognitive load. It also continues to encourage separation of UI and domain knowledge.
```

Comment:
@djsmith85 I added your suggestion here
> I'd suggest separating by feature/functionality rather than product though. As they are essentially dependencies, we should decide to pull them once needed. For example the importers are in common but are only used in web and cli.

I made some arguments for/against it further down but I just wanted to talk a little bit about another way of looking at it: Coupling vs Cohesion.

Stuff that has strong cohesion should preferably have a boundary created around it, while stuff that only  interact tangentially can instead be loosely coupled using other mechanisms. The way I usually look at it is 

- Cohesion: Direct reference to another class inside of a private field
- Coupling: Indirect references to other implementations using e.g. interfaces or other ways of abstraction. Two coupled packages usually contain knowledge about each other but can handled (eg. compiled) separately. 

Ex: `const organizationId: string` vs `const organizationId = new OrganizationId()`

Now I'm arguing that we should try to keep dependencies between packages to a minimum to avoid things like circular dependencies or excessive dependency on abstractions inside of a common package.

Reply:
Closed and moved to https://github.com/bitwarden/contributing-docs/pull/12/